packages/block-editor/src/hooks/metadata.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+import { has } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+import { hasBlockSupport } from '@wordpress/blocks';
+
+/**
+ * Filters registered block settings, extending attributes to include `metadata` in blocks declaring section support.
+ *
+ * @param {Object} settings Original block settings.
+ *
+ * @return {Object} Filtered block settings.
+ */
+export function addAttribute( settings ) {
+	if ( hasBlockSupport( settings, '__experimentalSection', false ) ) {
+		// Allow blocks to specify their own metadata attribute definition with default value if needed.
+		if ( ! has( settings.attributes, [ 'metadata' ] ) ) {
+			settings.attributes = {
+				...settings.attributes,
+				metadata: {
```

Comment:
What would be the difference with the new `settings` attribute added in https://github.com/WordPress/gutenberg/pull/40547 and the `meta` attribute proposed here?

Reply:
The settings attribute is a global styles settings object like the presets of a block, and other settings you can define using theme.json. It is equivalent to the styles object. The metadata is for information specific to a block instance that is not part of global styles/theme.json like the name of a specific section.

Reply:
It is though a good question (whether these could merge or not). I'm not sure I have any preference right now. They do seem to have different purposes.

Reply:
I'm really worried that we overuse `settings` and `metadata` which already exist in Block API but play a different role.

We use the block metadata phrase when talking about everything that `block.json` can represent:
https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-metadata.md

For settings the best way to show the challenge is by showing the implementation:

https://github.com/WordPress/gutenberg/blob/fa456ec87bb6504f0c116ef4e759e74b8873a4ff/packages/block-editor/src/hooks/settings.js#L16-L23

In effect, `settings` is an attribute in the block settings' object during registration.

There are also existing blocks that contain similar types of attributes that don't represent the content or visual aspects of the block and are included as regular attributes:
- `lock` https://github.com/WordPress/gutenberg/blob/fa456ec87bb6504f0c116ef4e759e74b8873a4ff/packages/block-editor/src/hooks/lock.js#L20
- `allowedBlocks`, example https://github.com/WordPress/gutenberg/blob/fa456ec87bb6504f0c116ef4e759e74b8873a4ff/packages/block-library/src/cover/block.json#L71
- `templateLock`, example https://github.com/WordPress/gutenberg/blob/fa456ec87bb6504f0c116ef4e759e74b8873a4ff/packages/block-library/src/cover/block.json#L74

I agree that it doesn't have too much difference from the implementation perspective and it's perfectly fine to proceed as is with `settings` and `metdata` attributes. However, it might be challenging to document all the subtle differences knowing about all the ambiguity introduced.

Reply:
> It is though a good question (whether these could merge or not). I'm not sure I have any preference right now. They do seem to have different purposes.

Just wanted to say that this has been on my mind since yesterday as well. And not only is `metadata` an overloaded term (block type metadata, post meta, Redux action meta, etc.), it's also a potential magnet for any random data. @jorgefilipecosta and @youknowriad: besides `name`, what do we contemplate storing in it?

> In effect, settings is an attribute in the block settings' object during registration.

I agree, @gziolo, that `settings` is also a bit overloaded, but it's also a first-class name in `theme.json`, and `attributes.settings` is symmetric with `attributes.style`, so IMO we could keep it. Maybe we could make the code clearer with more specific naming — for instance:

```diff
-       if ( ! settings?.attributes?.settings ) {
-               settings.attributes = {
-                       ...settings.attributes,
+       if ( ! blockTypeSettings?.attributes?.settings ) {
+               blockTypeSettings.attributes = {
+                       ...blockTypeSettings.attributes,
                        settings: {
                                type: 'object',
                        },
```

Reply:
>  @jorgefilipecosta and @youknowriad: besides name, what do we contemplate storing in it?

The two certain things is "name" and somekind of "tag" or "area" or "keyword" or something like that to add "semantics" to the given section/block. (better surface related patterns...). I think "locking" could be moved in that same object.

Potentially other things can come later. This comment is relevant here https://github.com/WordPress/gutenberg/issues/34352#issuecomment-1116300263

Reply:
It's true that having all the existing block settings and the new `settings` and `metadata` can be confusing about what goes where. IMO the new `settings` attribute is targeted for the `theme.json` mechanism so we could rename that one to something more indicative about its settings?

Other than that the metadata seems just fine to me and we can always look to migrate some other props there in the future properly..

Reply:
> settings attribute is targeted for the theme.json mechanism so we could rename that one to something more indicative about its purpose/contents?


Theme.json has styles and settings it seems natural that blocks also have styles and settings attributes. We don't have any other attribute for settings.

Reply:
We should follow @mcsf and use blockTypeSettings as a name to make things clearer. This PR already does that,  I will update the settings hook.